petastorm/reader_impl/reader_v2.py

```diff
@@ -0,0 +1,320 @@
+        logger.debug('dataset_url: %s', dataset_url)
+        resolver = FilesystemResolver(dataset_url)
+        self._dataset = pq.ParquetDataset(resolver.parsed_dataset_url().path, filesystem=resolver.filesystem(),
+                                          validate_schema=False)
+
+        # Get a unischema stored in the dataset metadata.
+        stored_schema = dataset_metadata.get_schema(self._dataset)
+
+        # Make a schema view (a view is a Unischema containing only a subset of fields
+        # Will raise an exception if invalid schema fields are in schema_fields
+        fields = schema_fields if isinstance(schema_fields, collections.Iterable) else None
+        self.schema = stored_schema.create_schema_view(fields) if fields else stored_schema
+
+        # 2. Get a list of all groups
+        row_groups = dataset_metadata.load_row_groups(self._dataset)
+
+        # 3. Filter rowgroups
+        filtered_row_groups, worker_predicate = self._filter_row_groups(self._dataset, row_groups, predicate,
+                                                                        rowgroup_selector, training_partition,
+                                                                        num_training_partitions)
+
+        epoch_items = self._apply_row_drop_partition(filtered_row_groups, shuffle_options)
+
+        # 4. Launch a new thread running `worker_loop` function.
+        epochs_iterator = lambda: epoch_generator(epoch_items, num_epochs, shuffle_options.shuffle_row_groups)
+
+        self._results_queue = Queue(_OUTPUT_QUEUE_SIZE)
+
+        loader = RowGroupLoader(dataset_url, self.schema, self.ngram, cache, worker_predicate)
+        decoder = RowDecoder(self.schema, self.ngram)
+        self._loader_pool = loader_pool or ThreadPoolExecutor(5)
+        self._decoder_pool = decoder_pool or ThreadPoolExecutor(5)
+        self._stop_flow_manager_event = threading.Event()
+        stats = Counter()
+
+        if not shuffling_queue:
+            shuffling_queue = NoopShufflingBuffer()
+
+        self._flow_manager_thread = threading.Thread(target=worker_loop,
+                                                     args=(epochs_iterator, self._loader_pool, loader,
+                                                           self._decoder_pool,
+                                                           decoder,
+                                                           shuffling_queue, self._results_queue,
+                                                           self._stop_flow_manager_event, stats))
+        self._flow_manager_thread.daemon = True
+        self._flow_manager_thread.start()
+
+        self._read_timeout_s = read_timeout_s
+
+    def _apply_row_drop_partition(self, row_groups, shuffle_options):
+        items_to_ventilate = []
+        for row_group in row_groups:
+            for shuffle_row_drop_partition in range(shuffle_options.shuffle_row_drop_partitions):
```

Comment:
This is missing normalizing the number of shuffle drop partitions https://github.com/uber/petastorm/blob/master/petastorm/reader.py#L268